### PR TITLE
=rem #xxxx use a Map watchee -> watchers in RemoteWatcher

### PR DESCRIPTION
**Pull request by trobert**
This commit replace the set of  (Watchee, Watchers) by a Map of Watchee -> Set of Watchers (a multimap actually)
This aims to remove whole iteration on all watches in most cases.

Some more commit should follow with other optimizations, especially for 1 watchee, N watchers case.

### CODE REVIEW

**Comment by patriknw** [comment]
@trobert thanks, I will take a look, but I can say right away that I don't think we will do this change in 2.3.x, so you should open another pull request targeted to master

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala [hunk 1]
```diff
@@ -102,10 +102,8 @@ private[cluster] class ClusterRemoteWatcher(
    * by super RemoteWatcher.
    */
   def takeOverResponsibility(address: Address): Unit = {
-    watching foreach {
-      case (watchee, watcher) ⇒ if (watchee.path.address == address)
-        unwatchRemote(watchee, watcher)
-    }
+    watching.keys.filter(_.path.address == address).foreach(clearAllWatches)
```
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: this loops twice, the filter can be moved inside the foreach as an if statement

Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: it's mostly creation of the intermediate collection that is the problem

File: akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala [hunk 2]
```diff
@@ -193,11 +193,11 @@ private[akka] class RemoteWatcher(
 
       // also watch from self, to be able to cleanup on termination of the watchee
       context watch watchee
-      watching += ((watchee, self))
+      watching += watchee → watching.get(watchee).fold(Set(self)) { _ + self }
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala]: slightly more efficient to use `get(watchee) match {`

File: akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala [hunk 3]
```diff
@@ -287,15 +296,15 @@ private[akka] class RemoteWatcher(
...
-        if (wee.path.address == address) {
-          // this re-watch will result in a RewatchRemote message to this actor
-          // must be a special message to be able to detect if an UnwatchRemote comes in
-          // before the extra RewatchRemote, then the re-watch should be ignored
-          log.debug("Re-watch [{} -> {}]", wer, wee)
-          wee.sendSystemMessage(new Rewatch(wee, wer)) // ➡➡➡ NEVER SEND THE SAME SYSTEM MESSAGE OBJECT TO TWO ACTORS ⬅⬅⬅
-        }
+    for {
+      (wee: InternalActorRef, wers) ← watching if wee.path.address == address
+      wer ← wers if wer.isInstanceOf[InternalActorRef]
```
Comment (patriknw) [inline akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala]: I would make that instanceOf into a `match` inside the next block, then isInstanceOf-asInstanceOf is not needed

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala (hunk 1 above)
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: also, I'm not sure, but I think it's more efficient to use `foreach { case (key, value)` than `keys`, because `keys` must create some kind of intermediate collection

**Comment by patriknw** [comment]
@trobert I like this, this part is not as scary as I first thought
anyway, open pull request to master, and we can decide later what/if backporting should take place

**Comment by trobert** [comment]
Thanks for the review!

I pushed another commit (the 1->N case), if you can have a look.

This a bit more intrusive, but the unit tests (and also my own testing) seems to be ok

I'll rebase against master and take your remark into account asap (probably next week). Is there any indicative schedule for akka 2.4 ?

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala (hunk 1 above)
Comment (trobert) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: keys is lightweight, it creates a wrapper around the map. You're right for filter however.
I propose a `for (watchee <- watching.keys`) with a guard condition (efficient and I find this a bit  more readable at the same time)

File: akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala (hunk 2 above)
Comment (trobert) [inline akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala]: Not sure about speed argument. `Option.fold` is marked inline and only consist of one `if`
But I change to a `match`, it is more common usage

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala (hunk 1 above)
Comment (ktoso) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: Or a plain `watching.keys.filterWith(_.path.address == address).foreach(clearAllWatches)` would jsut as well (exactly same as forloop).

File: akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala (hunk 3 above)
Comment (trobert) [inline akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala]: I propose

```
for {
      (wee, wers) ← watching if wee.path.address == address
      wer ← wers
      (watchee: InternalActorRef, watcher: InternalActorRef) = (wee, wer)
    }
```

File: akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala [hunk 4]
```diff
@@ -287,15 +287,12 @@ private[akka] class RemoteWatcher(
...
-        if (wee.path.address == address) {
-          // this re-watch will result in a RewatchRemote message to this actor
-          // must be a special message to be able to detect if an UnwatchRemote comes in
-          // before the extra RewatchRemote, then the re-watch should be ignored
-          log.debug("Re-watch [{} -> {}]", wer, wee)
-          wee.sendSystemMessage(new Rewatch(wee, wer)) // ➡➡➡ NEVER SEND THE SAME SYSTEM MESSAGE OBJECT TO TWO ACTORS ⬅⬅⬅
-        }
+    for {
+      wee ← watching.keys if wee.path.address == address
+      (watchee: InternalActorRef, watcher: InternalActorRef) = (wee, self)
```
Comment (ktoso) [inline akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala]: I'd avoid this pattern in performance sensitive code (`(a, b) = (x, y)`) - this allocates a Tuple2 each time (this may actually be the case where 2 Tuple2 allocations happen for each such line actually if I remember correctly).
A `match` inside would be allocation free and thus faster. :-)

**Comment by patriknw** [comment]
The second commit will require careful consideration, I will not be able to swap in that right now.
Thanks.

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala [hunk 5]
```diff
@@ -36,7 +36,7 @@ private[cluster] object ClusterRemoteWatcher {
  *
  * Specialization of [[akka.remote.RemoteWatcher]] that keeps
  * track of cluster member nodes and is responsible for watchees on cluster nodes.
- * [[akka.actor.AddressTerminate]] is published when node is removed from cluster.
+ * [[akka.actor.AddressTerminated]] is published when node is removed from cluster.
```
Comment (drewhk) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: nice!

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala [hunk 6]
```diff
@@ -73,8 +71,6 @@ private[cluster] class ClusterRemoteWatcher(
   override def receive = receiveClusterEvent orElse super.receive
 
   def receiveClusterEvent: Actor.Receive = {
-    case WatchRemote(watchee, watcher) if clusterNodes(watchee.path.address) ⇒
```
Comment (drewhk) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: hm, why is this removed?

File: akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala [hunk 7]
```diff
@@ -477,8 +461,13 @@ private[akka] class RemoteActorRef private[akka] (
 
   def sendSystemMessage(message: SystemMessage): Unit =
     try {
-      remote.send(message, None, this)
-      provider.afterSendSystemMessage(message)
+      message match {
+        case Watch(watchee, watcher) if watcher != provider.remoteWatcher ⇒
+          provider.remoteWatcher ! RemoteWatcher.WatchRemote(watchee, watcher)
```
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala]: wait, this change will now no longer send the Watch to the remote side

**Comment by drewhk** [comment]
OK TO TEST

File: akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala (hunk 7 above)
Comment (trobert) [inline akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala]: yes, this is the idea of the change: only watches from the remoteWatcher are sent to the remote side

File: akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala (hunk 6 above)
Comment (trobert) [inline akka-cluster/src/main/scala/akka/cluster/ClusterRemoteWatcher.scala]: `WatchRemote` message must be handled by the remote watcher for updating its watching map. This is necessary because remote watcher has now the responsibility to propagate the `DeathWatchNotifcation` to all watchers.

The heartbeat process will be bypassed for cluster members thanks to the override of the `watchNode` method

File: akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala (hunk 7 above)
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala]: so you say that remoteWatcher will send the Watch system message instead?

Comment (trobert) [inline akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala]: Exactly. this is to improve the case where different local actors are watching the same remote actor.
See discussion here: https://groups.google.com/d/msg/akka-dev/ao6MLKvxJ0o/INMBcgJ7PXMJ

**Comment by trobert** [comment]
The failure is linked to binary incompatibility. I'm opening a new PR against master as @patriknw suggested.

**Comment by patriknw** [comment]
yes, please

**Comment by patriknw** [comment]
Superseed by ##16911
